src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1085 @@
+                                        batch for description {:?}, in-flight: {:?}",
+                                    collection_id,
+                                    description,
+                                    in_flight_batches
+                                );
+                            }
+
+                            continue;
+                        }
+                        Event::Progress(frontier) => {
+                            batch_descriptions_frontier = frontier;
+                        }
+                    }
+                }
+                Some(event) = desired_input.next_mut() => {
+                    match event {
+                        Event::Data(_cap, data) => {
+                            // Extract desired rows as positive contributions to `correction`.
+                            if collection_id.is_user() && !data.is_empty() {
+                                trace!(
+                                    "persist_sink {collection_id}/{shard_id}: \
+                                        updates: {:?}, \
+                                        in-flight-batches: {:?}, \
+                                        desired_frontier: {:?}, \
+                                        batch_descriptions_frontier: {:?}",
+                                    data,
+                                    in_flight_batches,
+                                    desired_frontier,
+                                    batch_descriptions_frontier,
+                                );
+                            }
+
+                            // TODO: come up with a better default batch size here
+                            // (100 was chosen arbitrarily), and avoid having to make a batch
+                            // per-timestamp.
+                            for (row, ts, diff) in data.drain(..) {
+                                if write.upper().less_equal(&ts){
+                                    let builder = stashed_batches.entry(ts).or_insert_with(|| {
+                                        BatchBuilderAndMetdata::new(
+                                            write.builder(
+                                                // SUBTLE:
+                                                // The choice of the minimum antichain here
+                                                // over one that contains `ts` may seem odd, but
+                                                // currently, in persist, appended batches must
+                                                // be supersets of the of the `compare_and_append`
+                                                // description. This means we must choose a lower
+                                                // that is not greater than any batch description
+                                                // that may be grouped with this batch later on.
+                                                Antichain::from_elem(
```

Comment:
could we use something like `batch_descriptions_frontier` as a tighter lower bound?

Reply:
I don't think we can, but i can't remember lucidly why; ill think about it overnight

Reply:
This was fairly complex, but I added a commit on top that does this! I will be running nightly and RQ's against it!